.github/workflows/main.yml

```diff
@@ -36,59 +32,55 @@ jobs:
+          restore-keys: release-${{ runner.os }}-cargo-
+
       - name: Set environment variables
         run: |
           echo "PROJECT_NAME=$(cargo metadata --no-deps --format-version 1 | jq -r '.packages[0] | [ .name ] | join("")')" >> $GITHUB_ENV
           echo "PROJECT_VERSION=$(cargo metadata --no-deps --format-version 1 | jq -r '.packages[0] | [ .version ] | join("")')" >> $GITHUB_ENV
-      
+
       - name: Update dependencies
-        uses: actions-rs/cargo@v1
-        with:
-          command: update
+        run: cargo update
       - name: Build
-        uses: actions-rs/cargo@v1
-        with:
-          command: build
-          args: --release
+        run: cargo build --release
+      - name: Generate docs
+        run: cargo doc
 
-      # run tests, once normally, once to calculate code coverage
-      - name: Test
-        uses: actions-rs/cargo@v1
+      - name: Install cargo-tarpaulin
+        uses: baptiste0928/cargo-install@v2
         with:
-          command: test
-          args: --verbose
-      - name: Run tarpaulin to calculate code coverage
-        uses: actions-rs/tarpaulin@v0.1
-        with:
-          version: '0.15.0'
-          args: '-- --test-threads 1'
-          out-type: Html
-      
-      - name: Run clippy
-        uses: actions-rs/clippy-check@v1
+          crate: cargo-tarpaulin
+      - name: Calculate test coverage
+        run: cargo tarpaulin --out Html
+      - name: Archive code coverage results
+        uses: actions/upload-artifact@v1
         with:
-          token: ${{ secrets.GITHUB_TOKEN }}
-          args: -- -D warnings # escalate warnings s.t. pipeline fails
+          name: ${{ env.PROJECT_NAME }}-code_coverage_report-v${{ env.PROJECT_VERSION }}
+          path: tarpaulin-report.html
+
+      # Lints
+      - name: Clippy
+        run: cargo clippy -- -D warnings
       - name: Format
         run:  cargo fmt --all -- --check
       
       # Cargo check for security issues
-      - name: Security audit
-        uses: actions-rs/audit-check@v1
+      - name: Install cargo-audit
+        uses: baptiste0928/cargo-install@v2
         with:
-          token: ${{ secrets.GITHUB_TOKEN }}
-        
-      - name: Archive code coverage results
-        uses: actions/upload-artifact@v1
-        with:
-          name: ${{ env.PROJECT_NAME }}-code_coverage_report-v${{ env.PROJECT_VERSION }}
-          path: tarpaulin-report.html
+          crate: cargo-audit
+      - name: Security audit
+        run: cargo audit
+
+      # Check for outdated dependencies
+      - name: Install cargo-outdated
+        uses: dtolnay/install@cargo-outdated
+      - name: Outdated dependencies
+        run: cargo outdated --exit-code 1
+      
       - name: Archive release build
         uses: actions/upload-artifact@v1
         with:
           name: ${{ env.PROJECT_NAME }}-release_build-v${{ env.PROJECT_VERSION }}
           path: target/release/${{ env.PROJECT_NAME }}
-        
```

Comment:
Add newline back at end of file. (github has a warning for this when reviewing the changes)